### PR TITLE
*: support tls in restore operator

### PR DESCRIPTION
**Pull request by fanminshi**
restore operator now can created TLS enabled etcd cluster.

ref: https://github.com/coreos/etcd-operator/issues/1495

### CODE REVIEW

File: pkg/controller/restore-operator/sync.go [hunk 1]
```diff
@@ -136,9 +136,12 @@ func (r *Restore) prepareSeed(er *api.EtcdRestore) (err error) {
 	}()
 
 	cs := er.Spec.ClusterSpec
+	if err := cs.Validate(); err != nil {
+		return fmt.Errorf("invalid cluster spec. please fix the following problem with the cluster spec: %v", err)
```
Comment (xiang90) [inline pkg/controller/restore-operator/sync.go]: this is not a user facing error message. do not need to be this verbose. just invalid cluster spec: %v, error

File: test/e2e/e2eutil/spec_util.go [hunk 2]
```diff
@@ -88,6 +88,21 @@ func NewEtcdRestore(restoreName, version string, size int, restoreSource api.Res
 	}
 }
 
+// RestoreCRWithTLS attaches StaticTLS to its ClusterSpec.
+func RestoreCRWithTLS(er *api.EtcdRestore, memberPeerTLSSecret, memberServerTLSSecret, operatorClientTLSSecret string) {
+	if er != nil {
```
Comment (xiang90) [inline test/e2e/e2eutil/spec_util.go]:
```
if er == nil {
return
}
...
```

**Comment by fanminshi** [comment]
## Reproduce testing failure

Create namespace e2e:
`kubectl create ns e2e`

Create aws secret:

```sh
export NAMESPACE="e2e"
kubectl create secret generic aws --from-file=/Users/fanminshi/.aws/config --from-file=/Users/fanminshi/.aws/credentials -n $NAMESPACE
```

Run e2e TestBackupAndRestore: 

```sh
TEST_S3_BUCKET=jenkins-testing-operator TEST_AWS_SECRET=aws ./e2e.test -kubeconfig /Users/fanminshi/.kube/config -namespace $NAMESPACE -operator-image gcr.io/coreos-k8s-scale-testing/etcd-operator:fanmin -test.run TestBackupAndRestore
```

etcd operator log:

```
connecting to etcd endpoints: [https://restore-tls-test-8674665223082153551zhw2w-0000.restore-tls-test-8674665223082153551zhw2w.e2e.svc:2379]
time="2017-12-01T21:55:07Z" level=error msg="failed to update members: list members failed: creating etcd client failed: grpc: timed out when dialing" cluster-name=restore-tls-test-8674665223082153551zhw2w pkg=cluster
```

seed member log:

```
embed: rejected connection from "127.0.0.1:52114" (error "tls: failed to verify client's certificate: x509: certificate signed by unknown authority (possibly because of \"crypto/rsa: verification error\" while trying to verify candidate authority certificate \"Etcd Operator Example\")", ServerName "")
2017-12-01 21:55:02.411185 I | embed: rejected connection from "172.17.0.4:56510" (error "remote error: tls: bad certificate", ServerName "restore-tls-test-8674665223082153551zhw2w-0000.restore-tls-test-8674665223082153551zhw2w.e2e.svc")
2017-12-01 21:55:15.429952 I | embed: rejected connection from "172.17.0.4:56536" (error "remote error: tls: bad certificate", ServerName "restore-tls-test-8674665223082153551zhw2w-0000.restore-tls-test-8674665223082153551zhw2w.e2e.svc")
```

It seems to me that seed member is rejecting client request from etcd operator.

**Comment by fanminshi** [comment]
all fixed PTAL

**Comment by fanminshi** [comment]
all fixed PTAL

File: pkg/controller/restore-operator/sync.go [hunk 3]
```diff
@@ -136,9 +136,12 @@ func (r *Restore) prepareSeed(er *api.EtcdRestore) (err error) {
 	}()
 
 	cs := er.Spec.ClusterSpec
+	if err := cs.Validate(); err != nil {
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/sync.go]: This PR should be TLS support only.
Please put validating spec to a different PR/issue.

File: test/e2e/backup_test.go [hunk 4]
```diff
@@ -148,10 +148,19 @@ func testEtcdBackupOperatorForS3Backup(t *testing.T) string {
...
+	operatorClientTLSSecret := "restore-etcd-client-tls" + suffix
+	err := e2eutil.PrepareTLS(clusterName, f.Namespace, memberPeerTLSSecret, memberClientTLSSecret, operatorClientTLSSecret)
+	if err != nil {
+		t.Fatal(err)
+	}
 
 	restoreSource := api.RestoreSource{S3: e2eutil.NewS3RestoreSource(s3Path, os.Getenv("TEST_AWS_SECRET"))}
 	er := e2eutil.NewEtcdRestore("test-etcd-restore-", "3.2.11", 3, restoreSource)
-	er, err := f.CRClient.EtcdV1beta2().EtcdRestores(f.Namespace).Create(er)
+	er, err = f.CRClient.EtcdV1beta2().EtcdRestores(f.Namespace).Create(er)
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: TLS?

File: test/e2e/backup_test.go [hunk 5]
```diff
@@ -148,10 +148,19 @@ func testEtcdBackupOperatorForS3Backup(t *testing.T) string {
...
+	memberPeerTLSSecret := "restore-etcd-peer-tls" + suffix
+	memberClientTLSSecret := "restore-etcd-server-tls" + suffix
+	operatorClientTLSSecret := "restore-etcd-client-tls" + suffix
+	err := e2eutil.PrepareTLS(clusterName, f.Namespace, memberPeerTLSSecret, memberClientTLSSecret, operatorClientTLSSecret)
+	if err != nil {
+		t.Fatal(err)
+	}
 
 	restoreSource := api.RestoreSource{S3: e2eutil.NewS3RestoreSource(s3Path, os.Getenv("TEST_AWS_SECRET"))}
 	er := e2eutil.NewEtcdRestore("test-etcd-restore-", "3.2.11", 3, restoreSource)
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: One more question: why do we need to set version in NewEtcdRestore() instead of using default?

File: test/e2e/backup_test.go (hunk 4 above)
Comment (fanminshi) [reply test/e2e/backup_test.go]: ?

File: test/e2e/backup_test.go (hunk 5 above)
Comment (fanminshi) [reply test/e2e/backup_test.go]: was added in this pr https://github.com/coreos/etcd-operator/pull/1616.
cc/ @hasbro17

File: pkg/controller/restore-operator/sync.go (hunk 3 above)
Comment (fanminshi) [reply pkg/controller/restore-operator/sync.go]: sure

File: test/e2e/backup_test.go (hunk 4 above)
Comment (fanminshi) [reply test/e2e/backup_test.go]: i see what u mean.

File: test/e2e/backup_test.go [hunk 6]
```diff
@@ -148,10 +148,21 @@ func testEtcdBackupOperatorForS3Backup(t *testing.T) string {
...
+	operatorClientTLSSecret := "restore-etcd-client-tls" + suffix
+	err := e2eutil.PrepareTLS(clusterName, f.Namespace, memberPeerTLSSecret, memberClientTLSSecret, operatorClientTLSSecret)
+	if err != nil {
+		t.Fatal(err)
+	}
 
 	restoreSource := api.RestoreSource{S3: e2eutil.NewS3RestoreSource(s3Path, os.Getenv("TEST_AWS_SECRET"))}
-	er := e2eutil.NewEtcdRestore("test-etcd-restore-", "3.2.11", 3, restoreSource)
-	er, err := f.CRClient.EtcdV1beta2().EtcdRestores(f.Namespace).Create(er)
+	er := e2eutil.NewEtcdRestore("", "3.2.11", 3, restoreSource)
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: Why adds version in the parameter?
If unnecessary, can we have this fixed in this PR?

File: test/e2e/backup_test.go [hunk 7]
```diff
@@ -148,10 +148,21 @@ func testEtcdBackupOperatorForS3Backup(t *testing.T) string {
...
 	f := framework.Global
+	suffix := fmt.Sprintf("-%d", rand.Uint64())
+	clusterName := "restore-tls-test" + suffix
+	memberPeerTLSSecret := "restore-etcd-peer-tls" + suffix
+	memberClientTLSSecret := "restore-etcd-server-tls" + suffix
+	operatorClientTLSSecret := "restore-etcd-client-tls" + suffix
+	err := e2eutil.PrepareTLS(clusterName, f.Namespace, memberPeerTLSSecret, memberClientTLSSecret, operatorClientTLSSecret)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: defer delete the secrets?
Also do the same thing for backup test?

File: test/e2e/backup_test.go (hunk 6 above)
Comment (fanminshi) [reply test/e2e/backup_test.go]: prefer another pr since it is not related this tls pr.

File: test/e2e/backup_test.go (hunk 7 above)
Comment (fanminshi) [reply test/e2e/backup_test.go]: sure thing.

File: test/e2e/backup_test.go (hunk 6 above)
Comment (hongchaodeng) [reply test/e2e/backup_test.go]: Sure.

**Comment by fanminshi** [comment]
all fixed ptal.

**Comment by hongchaodeng** [comment]
LGTM

Please follow up to fix:
- add validate() 
- remove version param in NewRestoreCR()
